robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java

```diff
@@ -684,14 +690,52 @@ private String createTempDir(String name) {
 
   // TODO move/replace this with packageManager
   @VisibleForTesting
-  static void registerBroadcastReceivers(Application application, AndroidManifest androidManifest) {
+  static void registerBroadcastReceivers(
+      Application application, AndroidManifest androidManifest, LoadedApk loadedApk) {
     for (BroadcastReceiverData receiver : androidManifest.getBroadcastReceivers()) {
       IntentFilter filter = new IntentFilter();
       for (String action : receiver.getActions()) {
         filter.addAction(action);
       }
       String receiverClassName = replaceLastDotWith$IfInnerStaticClass(receiver.getName());
-      application.registerReceiver((BroadcastReceiver) newInstanceOf(receiverClassName), filter);
+      BroadcastReceiver broadcastReceiver = null;
```

Comment:
There may be a slightly simpler solution. Many months ago I did a quick prototype, and this also seems to work. It just involves instantiating things manually.

```java
  /**
   * Instantiate a new broadcast receiver, potentially using the {@link
   * android.app.AppComponentFactory} defined in the manifest.
   */
  private static BroadcastReceiver newBroadcastReceiverP(
      String receiverClassName, ApplicationInfo applicationInfo) {
    String appComponentFactoryName = applicationInfo.appComponentFactory;

    AppComponentFactory appComponentFactory = null;
    if (appComponentFactoryName != null) {
      appComponentFactoryName = replaceLastDotWith$IfInnerStaticClass(appComponentFactoryName);
      appComponentFactory = (AppComponentFactory) newInstanceOf(appComponentFactoryName);
    }
    if (appComponentFactory == null) {
      appComponentFactory = new AppComponentFactory();
    }
    try {
      return appComponentFactory.instantiateReceiver(
          Shadow.class.getClassLoader(), receiverClassName, null);
    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
      return null;
    }
  }
  ```

Reply:
Oh, it looks better. I will update it with this method.

Reply:
I decide to keep the current implementation with initializing `AppComponentFactory` and storing it in `LoadedApk` to avoid unnecessary initialization by following AOSP's implementation.